code/modules/admin/data_input.dm

```diff
@@ -0,0 +1,217 @@
+/// @param custom_type_message 	If not null, set as the text for input type selection
+/// @return 			 		A data_input_result with the parsed input and the selected input type, or both null if we didn't get any data
+proc/input_data(list/allowed_types, client/user, custom_title = null, custom_message = null, default = null, custom_type_title = null, custom_type_message = null, default_type = null)
+	. = new /datum/data_input_result(null, null) //in case anything goes wrong, return this. Thus, to check for cancellation, check for a null output_type.
+
+	if (!isclient(user)) //attempt to recover
+		if (ismob(user))
+			var/mob/mob_user = user
+			user = mob_user.client
+		else if (istype(user, /datum/mind))
+			var/datum/mind/user_mind
+			user = user_mind.current?.client
+
+	if (!isclient(user)) //rip
+		stack_trace("Tried to input data with non-client thing [user] \ref[user] of type [user.type].")
+		return
+
+	if (!user.holder)
+		message_admins("Non-admin client [user.key] somehow tried to input some data. Huh?")
+		logTheThing("debug", user.mob, null, "somehow attempted to input data via the input_data proc.")
+		return
+
+	var/input = null 	// The input from the user- usually text, but might be a file or something.
+	var/selected_type = input(custom_type_title || "Which input type?", custom_type_message || "Input Type Selection", default_type) as null|anything in allowed_types //TODO make this a TGUI list once we can indicate defaults on those
+
+	if (!selected_type)
+		return
+
+	if (selected_type != default_type) //clear the default if we aren't using the suggested type
+		default = null
+
+	switch(selected_type)
+		if (DATA_INPUT_NUM)
+			input = input(custom_title || "Enter number:", custom_message, default) as null|num
+
+		if (DATA_INPUT_TYPE)
+			var/stub = input(custom_title || "Enter part of type:", custom_message) as null|text
+			if (!stub)
+				boutput(user, "<span class='alert'>Cancelled.</span>")
+				return
+			input = get_one_match(stub, /datum, use_concrete_types = FALSE, only_admin_spawnable = FALSE)
+
+		if (DATA_INPUT_COLOR)
+			input = input(custom_title || "Select color:", custom_message, default) as null|color
+
+		if (DATA_INPUT_TEXT)
+			input = input(custom_title || "Enter text:", custom_message, default) as null|message
+
+		if (DATA_INPUT_ICON)
+			input = input(custom_title || "Select icon:", custom_message) as null|icon
+
+		if (DATA_INPUT_BOOL)
+			//lines written by the utterly insane
+			input = alert(custom_title || "True or False?", custom_message + (!isnull(default) ? "(Default: [default ? "True" : "False"])" : null), "True", "False") == "True" ? TRUE : FALSE
+
+		if (DATA_INPUT_LIST)
+			//TODO uhhhhhhhh h
+
+		if (DATA_INPUT_FILE)
+			input = input(custom_title || "Select file:", custom_message) as null|file
+
+		if (DATA_INPUT_DIR)
+			input = input(custom_title || "Enter direction:", "Dir as text (e.g. North), case doesn't matter", default) as null|text
```

Comment:
seems kind of silly to ask the user to type out a string when you are only allowing 8 valid inputs

Reply:
good point- replaced with list selection